internal/mode/static/state/dataplane/configuration.go

```diff
@@ -43,27 +43,117 @@ func BuildConfiguration(
 
 	upstreams := buildUpstreams(ctx, g.Gateway.Listeners, serviceResolver)
 	httpServers, sslServers := buildServers(g, generator)
+	passthroughServers := buildPassthroughServers(g)
+	streamUpstreams := buildStreamUpstreams(ctx, g.Gateway.Listeners, serviceResolver)
 	backendGroups := buildBackendGroups(append(httpServers, sslServers...))
 	keyPairs := buildSSLKeyPairs(g.ReferencedSecrets, g.Gateway.Listeners)
 	certBundles := buildCertBundles(g.ReferencedCaCertConfigMaps, backendGroups)
 	telemetry := buildTelemetry(g)
 	baseHTTPConfig := buildBaseHTTPConfig(g)
 
 	config := Configuration{
-		HTTPServers:    httpServers,
-		SSLServers:     sslServers,
-		Upstreams:      upstreams,
-		BackendGroups:  backendGroups,
-		SSLKeyPairs:    keyPairs,
-		Version:        configVersion,
-		CertBundles:    certBundles,
-		Telemetry:      telemetry,
-		BaseHTTPConfig: baseHTTPConfig,
+		HTTPServers:           httpServers,
+		SSLServers:            sslServers,
+		TLSPassthroughServers: passthroughServers,
+		Upstreams:             upstreams,
+		StreamUpstreams:       streamUpstreams,
+		BackendGroups:         backendGroups,
+		SSLKeyPairs:           keyPairs,
+		Version:               configVersion,
+		CertBundles:           certBundles,
+		Telemetry:             telemetry,
+		BaseHTTPConfig:        baseHTTPConfig,
 	}
 
 	return config
 }
 
+// buildPassthroughServers builds TLSPassthroughServers from TLSRoutes attaches to listeners.
+func buildPassthroughServers(g *graph.Graph) []Layer4VirtualServer {
+	passthroughServers := []Layer4VirtualServer{}
+	for _, l := range g.Gateway.Listeners {
+		if !l.Valid {
+			continue
+		}
+		if l.Source.Protocol == v1.TLSProtocolType {
+			for _, r := range l.L4Routes {
+				if !r.Valid {
+					continue
+				}
+				for _, h := range r.Spec.Hostnames {
+					passthroughServers = append(passthroughServers, Layer4VirtualServer{
+						Hostname:     string(h),
+						UpstreamName: r.Spec.BackendRef.ServicePortReference(),
+						Port:         int32(l.Source.Port),
+					})
+				}
+			}
+		}
+	}
+
+	return passthroughServers
+}
+
+// buildStreamUpstreams builds all the upstreams to be in the nginx stream block.
```

Comment:
nit: let's try and keep nginx out of this package whenever possible
```suggestion
// buildStreamUpstreams builds all stream upstreams.
```